csharp/src/SeedLang/Runtime/HeapObject.cs

```diff
@@ -222,12 +270,20 @@ internal partial class HeapObject : IEquatable<HeapObject> {
       return intIndex < 0 ? length + intIndex : intIndex;
     }
 
-    private static string TupleToString(IReadOnlyList<Value> tuple) {
+    private static void CheckKey(Value key) {
+      if (!key.IsNil && !key.IsBoolean && !key.IsNumber && !key.IsString && !key.IsRange &&
+          !key.IsTuple) {
+        throw new DiagnosticException(SystemReporters.SeedRuntime, Severity.Fatal, "", null,
+                                      Message.RuntimeErrorUnhashableType);
+      }
+    }
+
+    private static string TupleToString(ImmutableArray<Value> tuple) {
       var sb = new StringBuilder();
       sb.Append('(');
-      for (int i = 0; i < tuple.Count; i++) {
+      for (int i = 0; i < tuple.Length; i++) {
```

Comment:
I guess ImutableArray can be passed into String.Join so that the following lines can be replaced with a single line:

```
sb.Append(String.Join(", ", tuple));
```

Reply:
Fixed.